python/mlc_chat/model/llama/llama_model.py

```diff
@@ -102,31 +104,41 @@ def __init__(self, config: LlamaConfig):
             bias=False,
         )
         self.o_proj = nn.Linear(self.num_q_heads * self.head_dim, config.hidden_size, bias=False)
+        # KV cache for single sequence
         self.k_cache = nn.KVCache(config.context_window_size, [self.num_kv_heads, self.head_dim])
         self.v_cache = nn.KVCache(config.context_window_size, [self.num_kv_heads, self.head_dim])
 
-    def forward(  # pylint: disable=too-many-locals
+    def forward(  # pylint: disable=too-many-locals,too-many-arguments
         self,
         hidden_states: Tensor,
-        attention_mask: Tensor,
-        total_seq_len: tir.Var,
+        attention_mask: Optional[Tensor],
+        total_seq_len: Optional[tir.Var],
+        paged_kv_cache: Optional[PagedKVCache],
+        layer_id: int,
     ):
         d, h_q, h_kv, t = self.head_dim, self.num_q_heads, self.num_kv_heads, total_seq_len
         b, s, _ = hidden_states.shape
-        assert b == 1, "Only support batch size 1 at this moment."
-        # Step 1. QKV Projection
+        # QKV Projection
         qkv = self.qkv_proj(hidden_states)
         qkv = op.reshape(qkv, (b, s, h_q + h_kv + h_kv, d))
-        # Step 2. Apply QK rotary embedding
-        q, k, v = op_ext.llama_rope(qkv, t, self.rope_theta, h_q, h_kv)
-        # Step 3. Query and update KVCache
-        self.k_cache.append(op.squeeze(k, axis=0))
-        self.v_cache.append(op.squeeze(v, axis=0))
-        k = self.k_cache.view(t)
-        v = self.v_cache.view(t)
-        # Step 4. Compute softmax(Q @ K^T / sqrt(d)) @ V
-        output = op_ext.attention(q, k, v, casual_mask=attention_mask)
-        # Step 5. Apply output projection
+
+        if paged_kv_cache is None:
+            # Single sequence attention.
+            assert attention_mask is not None
+            assert t is not None
```

Comment:
remove this assert - there's no reason to believe `t` is nullable

Reply:
Yes it's true that currently `t` will never be None in this branch (`paged_kv_cache is None`). My intention is to add a safety check here, since though `t` will never be None for `prefill` and `decode`, it is None for `prefill_batch` and `decode_batch` https://github.com/mlc-ai/mlc-llm/blob/2295999cdb25cac96e1565bb1860bfd358c26672/python/mlc_chat/model/llama/llama_model.py#L398-L430

For this reason, I changed all the type annotations to `Optional`, and added the assertion to ensure in the future we don't mistakenly pass in a None for `t`.

Reply:
Ah I probably overlooked the possibility where `total_seq_len` must be `None` for batched setting. Thanks for the clarification!